### PR TITLE
Implement unit conversion

### PR DESCRIPTION
**Pull request by brandonpollett**
This PR implements unit conversion by leveraging the Units.Net library.

### CODE REVIEW

File: Microsoft.HealthVault.Fhir.UnitTests/ToHealthVaultTests/ObservationToHealthVaultWeightTests.cs [hunk 1]
```diff
@@ -20,14 +20,20 @@ namespace Microsoft.HealthVault.Fhir.UnitTests.ToHealthVaultTests
     [TestClass]
     public class ObservationToHealthVaultWeightTests
     {
-        [TestMethod]
-        public void WeightToHealthVault_Successful()
+        private static Observation GetObservation(string fileName)
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir.UnitTests/ToHealthVaultTests/ObservationToHealthVaultWeightTests.cs]: place at the bottom .. private methods follow public methods ..

File: Microsoft.HealthVault.Fhir.UnitTests/ToHealthVaultTests/ObservationToHealthVaultWeightTests.cs [hunk 2]
```diff
@@ -20,14 +20,20 @@ namespace Microsoft.HealthVault.Fhir.UnitTests.ToHealthVaultTests
...
+            var json = SampleUtil.GetSampleContent(fileName);
 
             var fhirParser = new FhirJsonParser();
             var observation = fhirParser.Parse<Observation>(json);
-            
+            return observation;
+        }
+
+        [TestMethod]
+        public void WeightToHealthVault_Successful()
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir.UnitTests/ToHealthVaultTests/ObservationToHealthVaultWeightTests.cs]: When _ Then pattern

File: Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs [hunk 3]
```diff
@@ -68,22 +71,37 @@ public static ThingBase ToHealthVault(this Observation observation)
 
         internal static T GetThingValueFromQuantity<T>(Quantity value) where T : Measurement<double>, new()
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: arg name as quantityValue?

Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: when returning null it is usually goes with TryGetThing

File: Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs [hunk 4]
```diff
@@ -68,22 +71,37 @@ public static ThingBase ToHealthVault(this Observation observation)
...
-                    result.DisplayValue = new DisplayValue();
-                    result.DisplayValue.Value = (double)value.Value;
-                    result.DisplayValue.Units = value.Unit;
-                    result.DisplayValue.UnitsCode = value.Code;
-                    return result;
-                }
+                return null;
             }
+            
+            var unitConversion = UnitResolver.Instance.UnitConversions.FirstOrDefault(x => x.Code == value.Code);
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: code is a string ? .. if so .. is it better to use OrdinalIgnoreCase while comparing ?

File: Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs [hunk 5]
```diff
@@ -68,22 +71,37 @@ public static ThingBase ToHealthVault(this Observation observation)
...
 
-            return null;
+            double convertedValue;
+            if (unitConversion != null)
+            {
+                var unitEnum = Enum.Parse(Type.GetType($"UnitsNet.Units.{unitConversion.UnitsNetUnitEnum}, UnitsNet"), unitConversion.UnitsNetSource);
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: you want to do Enum.TryParse ?

File: Microsoft.HealthVault.Fhir/Units/UnitConversion.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Microsoft.HealthVault.Fhir.Units
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Units/UnitConversion.cs]: copyright license ..

File: Microsoft.HealthVault.Fhir/Units/UnitResolver.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections.Generic;
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Units/UnitResolver.cs]: copyright license

File: Microsoft.HealthVault.Fhir/Units/UnitResolver.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        private static UnitResolver _instance;
+
+        public List<UnitConversion> UnitConversions { get; set; }
+
+        private UnitResolver()
+        {
+            UnitConversions = JsonConvert.DeserializeObject<List<UnitConversion>>(File.ReadAllText(@"Data\UnitConversions.json"));
+        }
+
+        public static UnitResolver Instance
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Units/UnitResolver.cs]: This has disadvantage of multiple threads instantiating the instance .. here are more details ... https://msdn.microsoft.com/en-us/library/ff650316.aspx

File: Microsoft.HealthVault.Fhir/Data/UnitConversions.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
+[
+    {
+        "Code": "cm",
+        "UnitsNetType": "Length",
+        "UnitsNetUnitEnum": "LengthUnit",
+        "UnitsNetSource": "Centimeter",
+        "UnitsNetDestination": "Meters"
+
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Data/UnitConversions.json]: remove extra line ?

File: Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs (hunk 5 above)
Comment (gcorvera) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: Agreed, TryParse should be used here

---
In reply to: [127042699](https://github.com/Microsoft/healthvault-fhir-library/pull/7#discussion_r127042699) [](ancestors = 127042699)

File: Microsoft.HealthVault.Fhir/Units/UnitResolver.cs (hunk 8 above)
Comment (gcorvera) [inline Microsoft.HealthVault.Fhir/Units/UnitResolver.cs]: Agree, please implement locks on the instance to ensure it is thread safe. 

---
In reply to: [127044951](https://github.com/Microsoft/healthvault-fhir-library/pull/7#discussion_r127044951) [](ancestors = 127044951)

File: Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs [hunk 10]
```diff
@@ -68,22 +71,37 @@ public static ThingBase ToHealthVault(this Observation observation)
...
+                var unitType = Type.GetType($"UnitsNet.{unitConversion.UnitsNetType}, UnitsNet");
+                var destinationObject = unitType.GetMethod("From", new[] {typeof(double), unitEnum.GetType()}).Invoke(null, new[] {(double)value.Value, unitEnum});
```
Comment (gcorvera) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: >(double)value.Value [](start = 131, length = 19)

Can this potentially have a cast exception?

File: Microsoft.HealthVault.Fhir/Microsoft.HealthVault.Fhir.csproj [hunk 11]
```diff
@@ -6,9 +6,21 @@
...
   <ItemGroup>
+    <None Remove="Data\UnitConversions.json" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Content Include="Data\UnitConversions.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+
```
Comment (gcorvera) [inline Microsoft.HealthVault.Fhir/Microsoft.HealthVault.Fhir.csproj]: We should add this as embedded resources and not content.

Comment (brandonpollett) [inline Microsoft.HealthVault.Fhir/Microsoft.HealthVault.Fhir.csproj]: Talking to @gcorvera, we're going to leave this as a content file for now to allow end users to potentially extend.

File: Microsoft.HealthVault.Fhir/Units/UnitResolver.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+using System.Collections.Generic;
+using System.IO;
+using Newtonsoft.Json;
+
+namespace Microsoft.HealthVault.Fhir.Units
+{
+    internal class UnitResolver
+    {
+        private static volatile UnitResolver s_instance;
+        private static object _lockInstance = new object();
```
Comment (gcorvera) [inline Microsoft.HealthVault.Fhir/Units/UnitResolver.cs]: >_l [](start = 30, length = 2)

This should be s_lockInstance

File: Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs [hunk 13]
```diff
@@ -66,24 +69,48 @@ public static ThingBase ToHealthVault(this Observation observation)
...
-                    result.DisplayValue = new DisplayValue();
-                    result.DisplayValue.Value = (double)value.Value;
-                    result.DisplayValue.Units = value.Unit;
-                    result.DisplayValue.UnitsCode = value.Code;
-                    return result;
+                    unitEnum = Enum.Parse(Type.GetType($"UnitsNet.Units.{unitConversion.UnitsNetUnitEnum}, UnitsNet"), unitConversion.UnitsNetSource);
                 }
+                catch (ArgumentNullException e)
+                {
+                    throw new ArgumentNullException("UnitsNetUnitEnum specified not found.", e);
```
Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: let's not do this .. throwing the same exception won't really help ..

File: Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs [hunk 14]
```diff
@@ -66,24 +70,52 @@ public static ThingBase ToHealthVault(this Observation observation)
...
+                return null;
+            }
+            
+            var unitConversion = UnitResolver.Instance.UnitConversions.FirstOrDefault(x => x.Code.Equals(quantityValue.Code, StringComparison.Ordinal));
+
+            double convertedValue;
+            if (unitConversion != null)
+            {
+                object unitEnum;
+                try
```
Comment (brandonpollett) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: The goal of this is to check that the UnitsNetUnitEnum exists in the UnitsNet.Units namespace and that the UnitsNetSource is able to be parsed out of that namespace. The errors that get thrown if you mistype are the same as if you didn't supply the parameter.

Comment (suryaren) [inline Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs]: Ok .. thanks for the clarification .. but do you want to check separately than passing to Enum .. I prefer to first check if the type exist, throw an exception if not, then pass to Enum ? you may want to do the same for the following statement for UnitsNet ? ..

**Pull request merged**